src/Assistant.ts

```diff
@@ -299,14 +300,20 @@ export async function processAssistantMiddleware(
  * was received. Alias for `postMessage()`.
  * https://api.slack.com/methods/chat.postMessage
  */
-function createSay(args: AllAssistantMiddlewareArgs): SayFn {
+async function createSay(args: AllAssistantMiddlewareArgs): Promise<SayFn> {
   const { client, payload } = args;
-  const { channelId: channel, threadTs: thread_ts } = extractThreadInfo(payload);
+  const { channelId: channel, threadTs: thread_ts, context } = extractThreadInfo(payload);
+  const threadContext = context.channel_id ? context : await args.getThreadContext(args);
```

Comment:
Ran this by @seratch already, but since this is a factory, every time an Assistant event is received, this logic will run. Default `get` relies on the `DefaultThreadContextStore`, but if the developer has provided a custom `threadContextStore`, the method they've provided will be called instead – _if_ no context is available off of the event itself.

Reply:
Sorry, my comment might be a bit unclear but what I mean is this createSay method should avoid calling the get method if possible. As you pointed out, bolt-python's code did this but it is now improved: https://github.com/slackapi/bolt-python/pull/1186

Reply:
I haven't checked if this code perfectly works but what I mean is something like this:
```diff
diff --git a/src/Assistant.ts b/src/Assistant.ts
index aec9a1d..2719ad2 100644
--- a/src/Assistant.ts
+++ b/src/Assistant.ts
@@ -133,7 +133,7 @@ export class Assistant {
 
   private async processEvent(args: AllAssistantMiddlewareArgs): Promise<void> {
     const { payload } = args;
-    const assistantArgs = await enrichAssistantArgs(this.threadContextStore, args);
+    const assistantArgs = enrichAssistantArgs(this.threadContextStore, args);
     const assistantMiddleware = this.getAssistantMiddleware(payload);
     return processAssistantMiddleware(assistantArgs, assistantMiddleware);
   }
@@ -161,10 +161,10 @@ export class Assistant {
  *  events from continuing down the global middleware chain to subsequent listeners
  *  2. Adds assistant-specific utilities (i.e., helper methods)
  * */
-export async function enrichAssistantArgs(
+export function enrichAssistantArgs(
   threadContextStore: AssistantThreadContextStore,
   args: AllAssistantMiddlewareArgs<AssistantMiddlewareArgs>, // TODO: the type here states that these args already have the assistant utilities present? the type here needs likely changing.
-): Promise<AllAssistantMiddlewareArgs> {
+): AllAssistantMiddlewareArgs {
   const { next: _next, ...assistantArgs } = args;
   const preparedArgs = { ...(assistantArgs as Exclude<AllAssistantMiddlewareArgs<AssistantMiddlewareArgs>, 'next'>) };
 
@@ -172,7 +172,7 @@ export async function enrichAssistantArgs(
   preparedArgs.getThreadContext = () => threadContextStore.get(args);
   preparedArgs.saveThreadContext = () => threadContextStore.save(args);
 
-  preparedArgs.say = await createSay(preparedArgs);
+  preparedArgs.say = createSay(preparedArgs);
   preparedArgs.setStatus = createSetStatus(preparedArgs);
   preparedArgs.setSuggestedPrompts = createSetSuggestedPrompts(preparedArgs);
   preparedArgs.setTitle = createSetTitle(preparedArgs);
@@ -300,19 +300,21 @@ export async function processAssistantMiddleware(
  * was received. Alias for `postMessage()`.
  * https://api.slack.com/methods/chat.postMessage
  */
-async function createSay(args: AllAssistantMiddlewareArgs): Promise<SayFn> {
+function createSay(args: AllAssistantMiddlewareArgs): SayFn {
   const { client, payload } = args;
   const { channelId: channel, threadTs: thread_ts, context } = extractThreadInfo(payload);
-  const threadContext = context.channel_id ? context : await args.getThreadContext(args);
 
-  return (message: Parameters<SayFn>[0]) => {
+  return async (message: Parameters<SayFn>[0]) => {
     const postMessageArgument: ChatPostMessageArguments =
       typeof message === 'string' ? { text: message, channel, thread_ts } : { ...message, channel, thread_ts };
+    const threadContext = context.channel_id ? context : await args.getThreadContext(args);
 
-    postMessageArgument.metadata = {
-      event_type: 'assistant_thread_context',
-      event_payload: threadContext as MessageMetadataEventPayloadObject,
-    };
+    if (threadContext) {
+      postMessageArgument.metadata = {
+        event_type: 'assistant_thread_context',
+        event_payload: threadContext as MessageMetadataEventPayloadObject,
+      };
+    }
 
     return client.chat.postMessage(postMessageArgument);
   };
```

Reply:
It seems like the idea is to collect the thread context just-in-time right when the developer calls `say`, rather than at the moment the event is delivered to the app. Is that right?

Reply:
@filmaj Yes, your understanding is correct.

Reply:
Got it. 🫡  Adjusted